src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -4399,48 +4399,94 @@ protected int notValidUtf8(Object string, int byteIndex,
     }
 
     @Primitive(name = "string_character_index", lowerFixnum = 2)
-    public abstract static class StringCharacterIndexPrimitiveNode extends PrimitiveArrayArgumentsNode {
+    @NodeChild(value = "string", type = RubyNode.class)
+    @NodeChild(value = "pattern", type = RubyNode.class)
+    @NodeChild(value = "offset", type = RubyNode.class)
+    public abstract static class StringCharacterIndexPrimitiveNode extends PrimitiveNode {
 
-        @TruffleBoundary
-        @Specialization
-        protected Object stringCharacterIndex(Object string, Object pattern, int offset,
-                @CachedLibrary(limit = "2") RubyStringLibrary stringLibrary,
-                @CachedLibrary(limit = "2") RubyStringLibrary patternLibrary,
-                @Cached RopeNodes.CalculateCharacterLengthNode calculateCharacterLengthNode) {
-            if (offset < 0) {
-                return nil;
-            }
+        @Child SingleByteOptimizableNode singleByteOptimizableNode = SingleByteOptimizableNode.create();
 
-            final Rope stringRope = stringLibrary.getRope(string);
-            final Rope patternRope = patternLibrary.getRope(pattern);
+        @CreateCast("string")
+        protected RubyNode coerceStringToRope(RubyNode string) {
+            return ToRopeNodeGen.create(string);
+        }
 
-            final int total = stringRope.byteLength();
-            int p = 0;
-            final int e = p + total;
+        @CreateCast("pattern")
+        protected RubyNode coercePatternToRope(RubyNode pattern) {
+            return ToRopeNodeGen.create(pattern);
+        }
+
+        @Specialization(guards = "offset < 0")
```

Comment:
I would think the given offset is never negative, but something we can clean up later.